server/src/main/java/org/opensearch/search/backpressure/settings/SearchShardTaskSettings.java

```diff
@@ -39,9 +93,99 @@ private static class Defaults {
+        "search_backpressure.search_shard_task.heap_variance",
+        Defaults.HEAP_VARIANCE_THRESHOLD,
+        0.0,
+        Setting.Property.Dynamic,
+        Setting.Property.NodeScope
+    );
+
+    /**
+     * Defines the window size to calculate the moving average of heap usage of completed search shard tasks.
+     */
+    private volatile int heapMovingAverageWindowSize;
+    public static final Setting<Integer> SETTING_HEAP_MOVING_AVERAGE_WINDOW_SIZE = Setting.intSetting(
+        "search_backpressure.search_shard_task.heap_moving_average_window_size",
+        Defaults.HEAP_MOVING_AVERAGE_WINDOW_SIZE,
+        0,
+        Setting.Property.Dynamic,
+        Setting.Property.NodeScope
+    );
+
     public SearchShardTaskSettings(Settings settings, ClusterSettings clusterSettings) {
         totalHeapPercentThreshold = SETTING_TOTAL_HEAP_PERCENT_THRESHOLD.get(settings);
+        this.cpuTimeMillisThreshold = SETTING_CPU_TIME_MILLIS_THRESHOLD.get(settings);
+        this.elapsedTimeMillisThreshold = SETTING_ELAPSED_TIME_MILLIS_THRESHOLD.get(settings);
+        this.heapPercentThreshold = SETTING_HEAP_PERCENT_THRESHOLD.get(settings);
+        this.heapVarianceThreshold = SETTING_HEAP_VARIANCE_THRESHOLD.get(settings);
+        this.heapMovingAverageWindowSize = SETTING_HEAP_MOVING_AVERAGE_WINDOW_SIZE.get(settings);
+        this.cancellationRatio = SETTING_CANCELLATION_RATIO.get(settings);
+        this.cancellationRate = SETTING_CANCELLATION_RATE.get(settings);
+        this.cancellationBurst = SETTING_CANCELLATION_BURST.get(settings);
+
         clusterSettings.addSettingsUpdateConsumer(SETTING_TOTAL_HEAP_PERCENT_THRESHOLD, this::setTotalHeapPercentThreshold);
+        clusterSettings.addSettingsUpdateConsumer(SETTING_CPU_TIME_MILLIS_THRESHOLD, this::setCpuTimeMillisThreshold);
+        clusterSettings.addSettingsUpdateConsumer(SETTING_ELAPSED_TIME_MILLIS_THRESHOLD, this::setElapsedTimeMillisThreshold);
+        clusterSettings.addSettingsUpdateConsumer(SETTING_HEAP_PERCENT_THRESHOLD, this::setHeapPercentThreshold);
+        clusterSettings.addSettingsUpdateConsumer(SETTING_HEAP_VARIANCE_THRESHOLD, this::setHeapVarianceThreshold);
+        clusterSettings.addSettingsUpdateConsumer(SETTING_HEAP_MOVING_AVERAGE_WINDOW_SIZE, this::setHeapMovingAverageWindowSize);
+        clusterSettings.addSettingsUpdateConsumer(SETTING_CANCELLATION_RATIO, this::setCancellationRatio);
+        clusterSettings.addSettingsUpdateConsumer(SETTING_CANCELLATION_RATE, this::setCancellationRate);
+        clusterSettings.addSettingsUpdateConsumer(SETTING_CANCELLATION_BURST, this::setCancellationBurst);
+    }
+
+    /**
+     * Callback listeners.
+     */
+    public interface Listener {
+        void onCancellationRatioSearchShardTaskChanged();
+
+        void onCancellationRateSearchShardTaskChanged();
+
+        void onCancellationBurstSearchShardTaskChanged();
     }
```

Comment:
1. rename to `SearchShardTaskCancellationListener`
2. rename individual methods to `onXChanged`

Reply:
We have this same listener in `SearchTask` and `SearchShardTask`. Extracted out this listener in a separate interface `CancellationListener`, which will be implemented by `SearchBackpressureState` to handle `onXChanged`, where X in {Ratio, Rate, Burst}.